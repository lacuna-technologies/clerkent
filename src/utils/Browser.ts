import { browser } from "webextension-polyfill-ts"
import Memoize from 'memoizee'

const getExtensionLink = Memoize(
  (string = ``) => browser.runtime.getURL(string),
)
const isFirefox = () => getExtensionLink().startsWith(`moz-extension://`)
const isChrome = () => getExtensionLink().startsWith(`chrome-extension://`)

const Browser = {
  getExtensionLink,
  isChrome,
  isFirefox,
}

export default Browser
