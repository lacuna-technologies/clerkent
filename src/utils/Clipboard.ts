import { browser } from 'webextension-polyfill-ts'

const readText = async () => {
  const permissionsRequest = await browser.permissions.request({
    permissions: [`clipboardRead`],
  })
  if(permissionsRequest){
    return navigator.clipboard.readText()
  }
  return ``
}

const isURI = (string: string) => (new RegExp(/.*(:\/\/|about:|chrome:).*/, `gi`)).test(string)
const hasTooLongWord = (string: string) => {
  const matches = string.match(/\w+/g)
  if(matches === null){
    return false
  }
  return matches.sort(
    (a, b) => b.length - a.length,
  )[0].length > 30
}
const isTooLong = (string: string) => string.length > 300
const hasInvalidCharacters = (string: string) => (new RegExp(/.*[<>\\{|}].*/, `gi`)).test(string)

const getPopupSearchText = async (): Promise<string> => {
  const clipboardText = await readText()
  const isValid = (
    clipboardText && clipboardText.length > 0 &&
    !isTooLong(clipboardText) && !isURI(clipboardText) &&
    !hasTooLongWord(clipboardText) && !hasInvalidCharacters(clipboardText)
  )
  if(isValid){
    return clipboardText
  }
  return ``
}

const Clipboard = {
  getPopupSearchText,
  readText,
}

export default Clipboard