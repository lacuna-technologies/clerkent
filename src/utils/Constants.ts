import type Law from '../types/Law'

const JURISDICTIONS = {
  AU: {
    emoji: `ðŸ‡¦ðŸ‡º`,
    id: `AU` as Law.JursidictionCode,
    name: `Australia`,
  },
  CA: {
    emoji: `ðŸ‡¨ðŸ‡¦`,
    id: `CA` as Law.JursidictionCode,
    name: `Canada`,
  },
  EU: {
    emoji: `ðŸ‡ªðŸ‡º`,
    id: `EU` as Law.JursidictionCode,
    name: `European Union`,
  },
  HK: {
    emoji: `ðŸ‡­ðŸ‡°`,
    id: `HK` as Law.JursidictionCode,
    name: `Hong Kong`,
  },
  MY: {
    emoji: `ðŸ‡²ðŸ‡¾`,
    id: `MY` as Law.JursidictionCode,
    name: `Malaysia`,
  },
  NZ: {
    emoji: `ðŸ‡³ðŸ‡¿`,
    id: `NZ` as Law.JursidictionCode,
    name: `New Zealand`,
  },
  SG: {
    emoji: `ðŸ‡¸ðŸ‡¬`,
    id: `SG` as Law.JursidictionCode,
    name: `Singapore`,
  },
  UK: {
    emoji: `ðŸ‡¬ðŸ‡§`,
    id: `UK` as Law.JursidictionCode,
    name: `United Kingdom`,
  },
}


const SG_DATABASES = {
  lawnet: {
    icon: ``,
    name: `Lawnet`,
    url: `https://www.lawnet.sg/lawnet/web/lawnet/home`,
  },
  lexis: {
    icon: ``,
    name: `Lexread`,
    url: `https://www.lexread.lexisnexis.com/`,
  },
  sc: {
    icon: ``,
    name: `Supreme Court`,
    url: `https://www.supremecourt.gov.sg/news/supreme-court-judgments`,
  },
  slw: {
    icon: ``,
    name: `Singapore Law Watch`,
    url: `https://www.singaporelawwatch.sg/Judgments`,
  },
  sso: {
    icon: ``,
    name: `Singapore Statutes Online`,
    url: `https://sso.agc.gov.sg/`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw Asia`,
    url: `https://www.westlawasia.com/singapore/`,
  },
}

const UK_DATABASES = {
  bailii: {
    icon: ``,
    name: `BAILII`,
    url: `https://bailii.org/`,
  },
  legislation: {
    icon: ``,
    name: `legislation.gov.uk`,
    url: `https://www.legislation.gov.uk/`,
  },
  lexis: {
    icon: ``,
    name: `LexisNexis`,
    url: `http://www.lexisnexis.com/uk/legal`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw`,
    url: `https://uk.westlaw.com/Browse/Home/WestlawUK`,
  },
}

const EU_DATABASES = {
  curia: {
    icon: ``,
    name: `CURIA`,
    url: `https://curia.europa.eu/juris/recherche.jsf?language=en`,
  },
  epo: {
    icon: ``,
    name: `EPO`,
    url: `https://www.epo.org/law-practice/case-law-appeals`,
  },
}

const HK_DATABASES = {
  hkliihk: {
    icon: ``,
    name: `HKLII.hk`,
    url: `https://www.hklii.hk/`,
  },
  hkliiorg: {
    icon: ``,
    name: `HKLII.org`,
    url: `https://www.hklii.org/eng/`,
  },
}

const CA_DATABASES = {
  canlii: {
    icon: ``,
    name: `CanLII`,
    url: `https://www.canlii.org/en`,
  },
}

const AU_DATABASES = {
  austlii: {
    icon: ``,
    name: `AustLII`,
    url: `https://www8.austlii.edu.au`,
  },
}

const MISC_DATABASES = {
  commonlii: {
    icon: ``,
    name: `CommonLII`,
    url: `http://www.commonlii.org`,
  },
}

const dedupeObjects = (inputObject: Record<string, Record<string, string>>, jurisdictionCode: Law.JursidictionCode) =>
  Object.entries(inputObject)
  .reduce((accumulator, [id, object]: [string, Record<string, string>]) => ({
    ...accumulator,
    [`${jurisdictionCode}_${id}`]: object,
  }), {})

const DATABASES: Record<string, Law.Database> = {
  ...dedupeObjects(SG_DATABASES, `SG`),
  ...dedupeObjects(UK_DATABASES, `UK`),
  ...dedupeObjects(EU_DATABASES, `EU`),
  ...dedupeObjects(HK_DATABASES, `HK`),
  ...dedupeObjects(CA_DATABASES, `CA`),
  ...dedupeObjects(AU_DATABASES, `AU`),
  ...MISC_DATABASES,
}

const SG_COURTS = {
  sgca: {
    id: `SGCA`,
    name: `Court of Appeal`,
  },
  sghc: {
    id: `SGHC`,
    name: `High Court`,
  },
}

const EU_COURTS = {
  cjeu: {
    id: `CJEU`,
    name: `Court of Justice of the European Union`,
  },
  epo: {
    id: `EPO`,
    name: `European Patent Office`,
  },
}

const COURTS: Record<string, Record<string, string>> = {
  ...dedupeObjects(SG_COURTS, `SG`),
  ...dedupeObjects(EU_COURTS, `EU`),
}

const Constants = {
  COURTS,
  DATABASES,
  JURISDICTIONS,
}

export default Constants