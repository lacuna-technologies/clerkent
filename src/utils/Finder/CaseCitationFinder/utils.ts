import Logger from "utils/Logger"

export const formatAbbr = (
  { abbr, appendum = null }:
  { abbr: string, appendum?: string },
) => {
  const end = appendum || ``
  return `${abbr
    .split(``)
    .map((letter: string) =>
      /[a-z]/i.test(letter)
        ? `${letter}\\.?`
        : letter,
    ).join(``)
    }${end}`
}

export const formatAbbrs = (
  abbrArray: { abbr: string, appendum?: string }[],
) => abbrArray.map(
  ({ abbr, appendum }) => formatAbbr({ abbr, appendum}),
).join(`|`)

export const cleanVolume = (volumeString: string) => volumeString.replace(/\./g, ``).toLowerCase().trim()

export const sortCitationsByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: string[],
): string[] => {
  const lastIfNotFound = (index: number) => (
    index === -1
      ? citationsArray.length + 1
      : index
  )
  return citationsArray.sort((a, b) => {
    const indexA = lastIfNotFound(
      abbrsList.findIndex(
        currentAbbr => new RegExp(
          formatAbbr(currentAbbr),
          `i`,
        ).test(a),
      ),
    )
    const indexB = lastIfNotFound(
      abbrsList.findIndex(currentAbbr => 
        new RegExp(formatAbbr(currentAbbr), `i`).test(b),
      ),
    )
    return indexA - indexB
  })
}

export const sortCasesByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: Law.Case[],
  attribute = `citation`,
): Law.Case[] => sortCitationsByVolume(
    abbrsList,
    (citationsArray).map(c => c[attribute]),
  ).map(c => (citationsArray).find(v => v[attribute] === c))