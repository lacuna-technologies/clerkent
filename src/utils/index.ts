export { default as Constants } from './Constants'
export { default as Finder } from './Finder'
export { default as Messenger } from './Messenger'
export { default as Request } from './Request'
export { default as Storage } from './Storage'
export { default as Logger } from './Logger'
export { default as Helpers } from './Helpers'
export { default as Clipboard } from './Clipboard'