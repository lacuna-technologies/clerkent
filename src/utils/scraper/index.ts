import SG from './SG'
import EW from './EW'

const scraper = {
  EW,
  SG,
}

export default scraper