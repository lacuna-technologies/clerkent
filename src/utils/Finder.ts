import Constants from './Constants'
import type Law from '../types/Law'
export interface FinderResult {
  jurisdiction: Law.JursidictionCode
  citation: string,
  index: number,
  court? : string
}

// const emptyParseResult: ParseResult = {
//   citation: undefined,
//   jurisdiction: undefined,
// }

// const parseQuery = (query: string): ParseResult => {
//   const cleanedQuery = query.trim()
//   return parseSGCase(cleanedQuery) || emptyParseResult
// }

// const inSLW = (query: string) => {
//   const eligibleCourt = Array.isArray(query.match(/SGCA|SGHC/))

//   const yearRegex = /^\[(2\d{3})]/
//   const yearMatch = query.match(yearRegex)
//   const eligibleYear = Array.isArray(yearMatch) && Number.parseInt(yearMatch[1]) >= 2000

//   return eligibleCourt && eligibleYear
// }

// const inSGSC = (query: string) => inSLW(query)

const findCase = (query: string): FinderResult[] => {
  return [
    ...findSGCase(query),
    ...findUKCase(query),
    ...findEUCase(query),
    ...findHKCase(query),
  ]
}

const findSGCase = (query: string): FinderResult[] => {
  const regex = /\[[12]\d{3}]( \d{1,2})? (sgca|sghc|sgdc|sgmc|slr(\(r\))?) \d{1,4}/gi
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
    }))
  }
  return []
}

const findUKCase = (query:string): FinderResult[] => {
  const abbrs = [
    {
      abbr: `EWCA`,
      appendum: `( Civ)?`,
    },
    {
      abbr: `EWHC`,
      appendum: `( Patents)?`,
    },
    {
      abbr: `UKSC`,
    },
    {
      abbr: `UKPC`,
    },
    {
      abbr: `UKHL`,
    },
    {
      abbr: `AC`,
    },
    {
      abbr: `UKHL`,
    },
    {
      abbr: `Ch`,
      appendum: `( D)?`,
    },
    {
      abbr: `QB`,
      appendum: `(D)?`,
    },
    {
      abbr: `KB`,
    },
    {
      abbr: `WLR`,
      appendum: `( \\(D\\))?`,
    },
    {
      abbr: `All ER`,
      appendum: `( \\(D\\))?`,
    },
    {
      abbr: `BCLC`,
    },
    {
      abbr: `BCC`,
    },
    {
      abbr: `HL Cas`,
    },
    {
      abbr: `App Cas`,
    },
    {
      abbr: `Ld Raym`,
    },
    {
      abbr: `FSR`,
    },
    {
      abbr: `ECC`,
    },
    {
      abbr: `ITCLR`,
    },
    {
      abbr: `RPC`,
    },
    {
      abbr: `Ex Rep`,
    },
    {
      abbr: `ER`,
    },
  ].map(({ abbr, appendum }) => `${abbr
    .split(``)
    .map(letter =>
      /[a-z]/i.test(letter)
        ? letter+`\\.?`
        : letter,
    ).join(``)
    }${appendum ? appendum : ``}`,
  ).join(`|`)
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/((\[|\()[12]\d{3}(-[12]\d{3})?(\]|\)))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.UK.id,
    }))
  }
  return []
}

export const epoRegex = new RegExp(/T ?\d{1,4}\/\d{1,2}/)
export const cjeuRegex = new RegExp(/[CT]-\d{1,3}\/\d{1,2}/)

const findEUCase = (query: string): FinderResult[] => {
  const regex = new RegExp(`(${epoRegex.source})|(${cjeuRegex.source})`, `gi`)
  const cleanedQuery = query.replaceAll(`â€‘`, `-`)

  const matches = [...cleanedQuery.matchAll(regex)]
  if(matches.length > 0){
    return matches.map((match) => ({
      citation: match[0],
      court: match[1] ? Constants.COURTS.EU_epo.id : Constants.COURTS.EU_cjeu.id,
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.EU.id,
    }))
  }
  return []
}

const findHKCase = (query: string): FinderResult[] => {
  const regex = /\[[12]\d{3}] (HKCA|HKCFA|HKCFI) \d{1,4}/g
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.HK.id,
    }))
  }
  return []
}

const Finder = {
  findCase,
  findEUCase,
  findHKCase,
  findSGCase,
  findUKCase,
}

export default Finder