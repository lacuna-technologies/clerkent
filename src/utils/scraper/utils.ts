import Fuse from 'fuse.js'
import Helpers from '../Helpers'
import Constants from 'utils/Constants'
import Storage from 'utils/Storage'
import Finder from "../Finder"

export const findCitation = (function_ = Finder.findCaseCitation, inputText: string) => {
  const results = function_(inputText)
  if(results.length > 0){
    return results[0].citation
  }
  return null
}

const longestCommonSubstring = (stringA: string, stringB: string) => {
  if (!stringA || !stringB) {
    return {
      length: 0,
      str: ``,
    }
  }

  let longestSubstr = ``
  let currentSubstr = ``
  for (const element of stringA) {
    currentSubstr += element
    if (!stringB.toLowerCase().includes(currentSubstr.toLowerCase())) {
      if (currentSubstr.length > longestSubstr.length) {
        longestSubstr = currentSubstr
      }
      currentSubstr = ``
    }
  }
  return currentSubstr.length > longestSubstr.length ? {
    length: currentSubstr.length,
    str: currentSubstr,
  } : {
    length: longestSubstr.length,
    str: longestSubstr,
  }
}

const probablySameCase = (caseNameA: string, caseNameB: string) => {
  const {
    length: longestSubstrLength,
  } = longestCommonSubstring(
    caseNameA,
    caseNameB,
  )
  const sameCaseProbability = longestSubstrLength / Math.min(caseNameA.length, caseNameB.length)
  return sameCaseProbability >= 0.8
}

export const sortByName = (query: string, cases: Law.Case[]) => {
  const cleanQuery = Helpers.removeCommonAppends(query.toLowerCase())
  const fuse = new Fuse(cases.map(({ name }) => Helpers.removeCommonAppends(name.toLowerCase())), { ignoreLocation: true })
  return fuse
    .search(cleanQuery)
    .map(({ refIndex }) => cases[refIndex])
}

export const databaseUse = async (jurisdictionId: Law.JurisdictionCode, databaseId: string, function_: () => Promise<Law.Case[]>): Promise<Law.Case[]> => {
  const databasesStatus: typeof Constants.DEFAULT_DATABASES_STATUS = await Storage.get(`DATABASES_STATUS`) || Constants.DEFAULT_DATABASES_STATUS
  if(databasesStatus[jurisdictionId][databaseId]){
    return function_()
  } 
  return []
}