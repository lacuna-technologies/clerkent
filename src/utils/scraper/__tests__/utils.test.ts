import { sortByName } from "../utils"

type TestCase = {
  cases: string[],
  index: number,
  query: string,
}

const testCases: TestCase[] = [
  {
    cases: [
      `RANGE CONSTRUCTION PTE LTD v GOLDBELL ENGINEERING PTE LTD`,
      `Re: Aathar Ah Kong Andrew`,
      `Aathar Ah Kong Andrew v CIMB Securities (Singapore) Pte Ltd`,
      `Azman bin Kamis v Saag Oilfield Engineering (S) Pte Ltd (formerly known as Derrick Services Singapore Pte Ltd) and another suit`,
      `Lee Tat Development Pte Ltd v Management Corporation of Grange Heights Strata Title No 301 (No 2)`,
    ],
    index: 0,
    query: `range`,
  },
  {
    cases: [
      `UJF v UJG`,
      `Azman bin Kamis v Saag Oilfield Engineering (S) Pte Ltd (formerly known as Derrick Services Singapore Pte Ltd) and another suit`,
      `Tribune Investment Trust Inc v Soosan Trading Co Ltd`,
      `Lee Tat Development Pte Ltd v Management Corporation of Grange Heights Strata Title No 301 (No 2)`,
    ],
    index: 2,
    query: `tribune investment v soosan`,
  },
  {
    cases: [
      `Chua Choon Lim Robert v MN Swami and Others`,
      `Chua Teck Chew Robert v Goh Eng Wah`,
      `KOH CHEW CHEE v LIU SHU MING & Anor`,
      `Goh Nellie v Goh Lian Teck and Others`,
    ],
    index: 1,
    query: `Chua Teck Chew Robert v Goh Eng Wah`,
  },
  {
    cases: [
      `Choong Peng Kong v Koh Hong Son`,
      `China Airlines Ltd v Philips Hong Kong Ltd`,
      `China Airlines Limited v Philips Hong Kong Limited`,
      `Kong Chee Chui and others v Soh Ghee Hong`,
      `CRRC (HONG KONG) CO. LIMITED & Anor v CHEN WEIPING`,
      `Fustar Chemicals Ltd (Hong Kong) v Liquidator of Fustar Chemicals Pte Ltd`,
      `Fustar Chemicals Ltd v Ong Soo Hwa (liquidator of Fustar Chemicals Pte Ltd)`,
    ],
    index: 5,
    query: `Fustar Chemicals (Hong Kong)`,
  },
  {
    cases: [
      `PEX International Pte Ltd v Lim Seng Chye & Anor`,
      `Tan Chi Min v The Royal Bank of Scotland Plc`,
      `WestLB AG v Philippine National Bank & Others`,
      `Cousins Scott William v The Royal Bank of Scotland plc`,
      `The Royal Bank of Scotland NV (formerly known as ABN Amro Bank NV) and others v TT International Ltd and another appeal`,
      `The Royal Bank of Scotland NV (formerly known as ABN Amro Bank NV) and others v TT International Ltd (nTan Corporate Advisory Pte Ltd and others, other parties) and another appeal`,
    ],
    index: 4,
    query: `Royal Bank of Scotland v TT International`,
  },
]

describe(`scraper utils`, () => {

  test.concurrent
    .each(testCases.map((test) => [test.query, test.cases, test.index]))(`sort case names: %s`, (query, cases, id) => {
    const sorted = sortByName(query, cases.map(c => ({ name: c }) as Law.Case))
    expect(sorted[0].name).toBe(cases[id])
  })
})