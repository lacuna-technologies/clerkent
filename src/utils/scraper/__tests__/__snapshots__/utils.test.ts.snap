// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scraper utils sorts case names correctly 1`] = `
Array [
  Object {
    "name": "RANGE CONSTRUCTION PTE LTD v GOLDBELL ENGINEERING PTE LTD",
  },
  Object {
    "name": "Lee Tat Development Pte Ltd v Management Corporation of Grange Heights Strata Title No 301 (No 2)",
  },
  Object {
    "name": "Re: Aathar Ah Kong Andrew",
  },
  Object {
    "name": "Aathar Ah Kong Andrew v CIMB Securities (Singapore) Pte Ltd",
  },
  Object {
    "name": "Azman bin Kamis v Saag Oilfield Engineering (S) Pte Ltd (formerly known as Derrick Services Singapore Pte Ltd) and another suit",
  },
]
`;

exports[`scraper utils sorts case names correctly 2`] = `
Array [
  Object {
    "name": "Tribune Investment Trust Inc v Soosan Trading Co Ltd",
  },
  Object {
    "name": "UJF v UJG",
  },
  Object {
    "name": "Lee Tat Development Pte Ltd v Management Corporation of Grange Heights Strata Title No 301 (No 2)",
  },
  Object {
    "name": "Azman bin Kamis v Saag Oilfield Engineering (S) Pte Ltd (formerly known as Derrick Services Singapore Pte Ltd) and another suit",
  },
]
`;
